// @sudoku/stores/candidateHint.js
import { writable } from 'svelte/store';

const EMPTY_HINT = {
	visible: false,
	ok: false,
	reason: null,
	message: null,
	row: null,
	col: null,
	candidates: [],
};

const AUTO_HIDE_DELAY = 3000;

function normalizeCoordinate(value) {
	return Number.isInteger(value) ? value : null;
}

function createCandidateHint() {
	const hint = writable({ ...EMPTY_HINT });
	let hideTimer = null;

	function clearHideTimer() {
		if (hideTimer !== null) {
			clearTimeout(hideTimer);
			hideTimer = null;
		}
	}

	return {
		subscribe: hint.subscribe,

		show(result, cursor = {}) {
			clearHideTimer();

			const resultRow = normalizeCoordinate(result.row);
			const resultCol = normalizeCoordinate(result.col);
			const row = resultRow === null ? normalizeCoordinate(cursor.y) : resultRow;
			const col = resultCol === null ? normalizeCoordinate(cursor.x) : resultCol;

			hint.set({
				visible: true,
				ok: result.ok,
				reason: result.reason || null,
				message: result.message,
				row,
				col,
				candidates: result.ok ? result.candidates.slice() : [],
			});

			hideTimer = setTimeout(() => {
				hint.set({ ...EMPTY_HINT });
				hideTimer = null;
			}, AUTO_HIDE_DELAY);
		},

		hide() {
			clearHideTimer();
			hint.set({ ...EMPTY_HINT });
		},
	};
}

// 功能：保存候选提示查询结果，供操作栏和棋盘格共同渲染中文提示。
export const candidateHint = createCandidateHint();
