// @sudoku/game.js
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku } from '@sudoku/sudoku';
import { createGame, createSudoku } from '../../domain/index.js';
import { cursor } from './stores/cursor';
import { difficulty } from './stores/difficulty';
import { gamePaused } from './stores/game';
import { gameSession } from './stores/grid';
import { timer } from './stores/timer';
import { hints } from './stores/hints';

/**
 * 功能：根据题面初始化一局新的前端游戏会话。
 * 上下文：新局和自定义题面加载都会通过这里创建 Sudoku / Game。
 *
 * @param {number[][]} puzzleGrid - 9x9 数独题面
 */
function startSession(puzzleGrid) {
	const puzzle = createSudoku(puzzleGrid);
	const sessionGame = createGame({ sudoku: puzzle });

	gameSession.loadSession({
		puzzle,
		game: sessionGame,
	});

	cursor.reset();
	timer.reset();
	hints.reset();
}

/**
 * 功能：使用随机生成题面开始新游戏。
 * 上下文：欢迎弹窗和难度菜单会调用此函数。
 *
 * @param {('veryeasy' | 'easy' | 'medium' | 'hard')} diff - Difficulty
 */
export function startNew(diff) {
	difficulty.set(diff);
	startSession(generateSudoku(diff));

	location.hash = '';
}

/**
 * 功能：使用分享码中的题面开始自定义游戏。
 * 上下文：该流程同样创建领域层 Game，而不是直接写旧二维数组。
 *
 * @param {string} sencode - Sencode to decode
 */
export function startCustom(sencode) {
	difficulty.setCustom();
	startSession(decodeSencode(sencode));
}

/**
 * 功能：撤销当前游戏的一步操作。
 * 上下文：操作栏 Undo 按钮调用此函数，实际逻辑委托给领域层 Game。
 *
 * @returns {boolean} true 表示成功撤销
 */
export function undoGame() {
	return gameSession.undo();
}

/**
 * 功能：重做当前游戏的一步操作。
 * 上下文：操作栏 Redo 按钮调用此函数，实际逻辑委托给领域层 Game。
 *
 * @returns {boolean} true 表示成功重做
 */
export function redoGame() {
	return gameSession.redo();
}

/**
 * 功能：查询当前是否可以撤销。
 * 上下文：用于控制 Undo 按钮状态。
 *
 * @returns {boolean} true 表示可撤销
 */
export function canUndo() {
	return gameSession.canUndo();
}

/**
 * 功能：查询当前是否可以重做。
 * 上下文：用于控制 Redo 按钮状态。
 *
 * @returns {boolean} true 表示可重做
 */
export function canRedo() {
	return gameSession.canRedo();
}

/**
 * 功能：查询指定格子的候选提示。
 * 上下文：UI Hint 按钮调用此函数；函数本身不修改棋盘、不进入历史。
 *
 * @param {number|null} row - 行索引
 * @param {number|null} col - 列索引
 * @returns {{ ok: boolean, reason?: string, message: string, row?: number, col?: number, candidates?: number[] }}
 */
export function getCandidateHint(row, col) {
	return gameSession.getCandidateHint(row, col);
}

/**
 * 功能：应用下一步提示。
 * 上下文：UI 下一步提示按钮调用此函数；成功时会写入棋盘并进入 Game history。
 *
 * @returns {{ ok: boolean, reason?: string, message: string, row?: number, col?: number, value?: number }}
 */
export function applyNextStepHint() {
	return gameSession.applyNextStepHint();
}

/**
 * 功能：进入探索模式。
 * 上下文：当没有唯一候选数时由探索弹窗调用。
 *
 * @returns {{ ok: boolean, message: string }}
 */
export function startExplore() {
	return gameSession.startExplore();
}

/**
 * 功能：退出探索模式并把探索 history 合并回主模式。
 * 上下文：探索弹窗中的退出按钮调用，保证普通模式可继续 Undo/Redo 探索路径。
 *
 * @returns {{ ok: boolean, message: string }}
 */
export function exitExplore() {
	return gameSession.exitExplore();
}

/**
 * 功能：一键回到探索开始前的局面，同时保留探索 history。
 * 上下文：探索弹窗中的回到起点按钮调用。
 *
 * @returns {{ ok: boolean, message: string }}
 */
export function returnToExploreStart() {
	return gameSession.returnToExploreStart();
}

/**
 * 功能：查询当前是否处于探索模式。
 *
 * @returns {boolean}
 */
export function isExploring() {
	return gameSession.isExploring();
}

/**
 * 功能：查询探索模式状态。
 *
 * @returns {{ active: boolean, index: number, length: number, returnedToStart: boolean }}
 */
export function getExploreStatus() {
	return gameSession.getExploreStatus();
}

/**
 * 功能：查询探索模式下当前格候选数，并触发失败记忆检查。
 *
 * @param {number|null} row - 行索引
 * @param {number|null} col - 列索引
 * @returns {{ ok: boolean, reason?: string, message: string, row?: number, col?: number, candidates?: number[] }}
 */
export function getExploreCandidateHint(row, col) {
	return gameSession.getExploreCandidateHint(row, col);
}

/**
 * 功能：消费领域对象暂存的探索提示。
 *
 * @returns {{ reason: string, message: string }|null}
 */
export function consumeExploreNotice() {
	return gameSession.consumeExploreNotice();
}

/**
 * Pause the game
 */
export function pauseGame() {
	timer.stop();
	gamePaused.set(true);
}

/**
 * Resume (un-pause) the game
 */
export function resumeGame() {
	timer.start();
	gamePaused.set(false);
}

export default {
	startNew,
	startCustom,
	undo: undoGame,
	redo: redoGame,
	canUndo,
	canRedo,
	getCandidateHint,
	applyNextStepHint,
	startExplore,
	exitExplore,
	returnToExploreStart,
	isExploring,
	getExploreStatus,
	getExploreCandidateHint,
	consumeExploreNotice,
	pause: pauseGame,
	resume: resumeGame
};
